src/Model/PhasorDynamics/SystemModel.hpp

```diff
@@ -124,12 +125,37 @@ namespace GridKit
         for (const auto& gendata : data.genrou)
         {
           IdxT bus_index = 0;
-          if (gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::bus))
+          IdxT pmech_index = 0;
+          IdxT speed_index = 0;
+
+          bool has_bus = gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::bus);
+          bool has_pmech_signal = gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::pmech_signal);
+          bool has_speed_signal = gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::speed_signal);
+
+          if (has_bus && has_pmech_signal && has_speed_signal)
           {
+
             bus_index = gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::bus);
+            pmech_index = gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::pmech_signal);
+            speed_index = gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::speed_signal);
+
+            auto* gen = new Genrou<ScalarT, IdxT>(
+              getBus(bus_index), 
+              getBus(pmech_index),
+              getBus(speed_index),
+              gendata);
+            addComponent(gen);
+          }
+          else{
+            bus_index = gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::bus);
+            
+            auto* gen = new Genrou<ScalarT, IdxT>(
+              getBus(bus_index), 
+              nullptr,
+              nullptr,
+              gendata);
+            addComponent(gen);
           }
```

Comment:
```suggestion
          bus_type* bus   = nullptr;
          bus_type* pmech = nullptr;
          bus_type* speed = nullptr;

          if (gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::bus))
          {
            bus = getBus(gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::bus));
          }

          if (gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::pmech_signal))
          {
            pmech = getBus(gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::pmech_signal));
          }

          if (gendata.ports.contains(GenrouData<ScalarT, IdxT>::Ports::speed_signal))
          {
            speed = getBus(gendata.ports.at(GenrouData<ScalarT, IdxT>::Ports::speed_signal));
          }

          auto* gen = new Genrou<ScalarT, IdxT>(bus, pmech, speed, gendata);
          addComponent(gen);
```
you can make this much cleaner